### PR TITLE
Registry sketch and implementation for EVM chains

### PR DESCRIPTION
**Pull request by tolak**
This PR introduces the `Registry` module, which will contain the registration of
- EVM chain registration (implemented in this PR)
- Substrate chain registration (upcoming PR)
- Bridge registration (upcoming PR)
- Dex registration (upcoming PR)

### CODE REVIEW

**Comment by tolak** [comment]
Convert this PR to draft, will be opened again when finished refactor

File: index/Cargo.toml [hunk 1]
```diff
@@ -15,6 +18,9 @@ hex = { version = "0.4.3", default-features = false }
 pink-extension = { version = "0.1", default-features = false, features = ["ink-as-dependency"] }
 pink-web3 = { git = "https://github.com/Phala-Network/pink-web3.git", branch = "pink", default-features = false, features = ["pink", "signing"]}
 xcm = { git = "https://github.com/paritytech/polkadot", branch = "release-v0.9.29", default-features = false }
+ink_storage = { version = "3", default-features = false }
```
Comment (tenheadedlion) [inline index/Cargo.toml:21]: In a previous PR I deliberately remove these ink! dependencies, is there a way we can do without them?

File: index/src/registry/res/erc20-abi.json [hunk 2]
```diff
@@ -0,0 +1,156 @@
+[
```
Comment (tenheadedlion) [inline index/src/registry/res/erc20-abi.json:1]: Can we move this `abi.json` into the contracts/ directory?

**Review by tenheadedlion — COMMENTED** [review]
LGTM, there are 2 points I want to discuss:

1. Is there a workaround for bringing `ink` dependencies to the `index` module? The reason I am not keen on `ink` libraries is that they lessen the feel of `SDK` in `index/`. 
2. abi.json should be placed in the application layer(the `contracts/`)

For the first point, I think it's possible to get rid of `SpreadLayout` and `PackedLayout`

```rust
// somewhere in index/
// don't care about SpreadLayout, which may not make it to ink4.0, introducing it breaks the abstraction
pub struct EvmChain {
    info: ChainInfo,
    assets: Vec<AssetInfo>
}

// somewhere in contracts/
#[ink(storage)]
pub struct SomeStorage {
   map: Mapping<String, Vec<u8>> // or [u8; 1024]
}

// in a contract method:
fn register_method(args: ...) {
  // create an EvmChain
 let chain = EvmChain::from(args...);
 self.map.insert(args.name, 
  chain.encode() // or: somehow, if `Vec<u8>` is not allowed as the value of the mapping, think of a way to do this: `chain.to_array()`
);
}
```

File: index/src/registry/res/erc20-abi.json (hunk 2 above)
Comment (tolak) [reply index/src/registry/res/erc20-abi.json:1]: If `index` contains stuff to interact with EVM contract, it's ok to put it here. Why you said it should belong to the application layer?

**Review by tenheadedlion — APPROVED** [review]
There are disagreements on styles and readability, let's find another time to talk about these things

**Pull request merged**